File: src/bench_hsi.R
library(stars)
library(dplyr)

FILE <- "data/aviris_stress.zarr"

# 1. OPEN (Lazy Proxy)
# stars proxy object points to the Zarr on disk without loading
cube <- read_stars(FILE, proxy = TRUE)

cat(sprintf("Dataset Loaded (Proxy)\n"))

start <- Sys.time()

# 2. COMPUTE
# st_apply automatically chunks operations on proxy objects
# MARGIN 2:3 means keep x/y (dims 2 and 3), collapse dim 1 (band)
# Note: Dimension names might depend on how Zarr was written.
# We assume standard order. If error, check print(cube) dimensions.
result <- st_apply(cube, MARGIN = c("dim_1", "dim_2"), FUN = mean)

# Force computation by pulling result into memory
res_mem <- as.data.frame(result)

elapsed <- as.numeric(Sys.time() - start)
cat(sprintf("HSI Streaming Mean: %.4fs\n", elapsed))

File: src/bench_hsi.py
import time
import xarray as xr
import dask.array as da

FILE = "data/aviris_stress.zarr"

def main():
    # 1. OPEN (Lazy)
    ds = xr.open_zarr(FILE)
    data = ds['reflectance']

    print(f"Dataset Shape: {data.shape}")
    print(f"Dataset Size:  {data.nbytes / 1e9:.2f} GB")

    # 2. COMPUTE (Streaming)
    start = time.time()
    # Collapse Band Dimension (axis 0)
    # Dask will pull chunks, compute mean, discard chunk, repeat.
    result = data.mean(dim='dim_0').compute()
    elapsed = time.time() - start

    print(f"HSI Streaming Mean: {elapsed:.4f}s")

if __name__ == "__main__":
    main()
